sophia-fungible-token-contract.md

```diff
@@ -0,0 +1,406 @@
+
+```
+public stateful function init() = {
+    _owner = Call.caller,
+    _totalSupply = 0,
+    _balances = {},
+    _allowed = {}}
+```
+
+For now, our Sophia contract code looks like this: 
+```
+contract FungibleToken =
+  record state = {
+    _owner            : address,
+    _totalSupply      : int,
+    _balances         : map(address, int),
+    _allowed          : map((address,address), int)}
+
+  public stateful function init() = {
+    _owner = Call.caller,
+    _totalSupply = 0,
+    _balances = {},
+    _allowed = {}}
+```
+
+### Helper functions
+
+We will add some private helper functions in order to make our contract easier to read:
+
+- **require** - evaluates the condition passed as first parameter and throws with the value of second parameter on false;
+```
+private function require(b : bool, err : string) =
+    if(!b)
+      abort(err)
+```
+
+- **add** - returns the sum of two integers, throws on overflow;
+```
+private function add(_a : int, _b : int) : int =
+    let c : int = _a + _b
+    require(c >= _a, "Error")
+    c
+```
+- **sub** - subtracts two unsigned integers and return the result, throws on overflow (i.e. if subtrahend is greater than minuend);
+```
+private function sub(_a : int, _b : int) : int =
+    require(_b =< _a, "Error")
+    _a - _b
+```
+
+- **onlyOwner** - throws if called by any account other than the owner
+```
+private function onlyOwner() =
+      require(Call.caller == state._owner, "Only owner can mint!")
+```
+
+- **lookupByAddress** - checks if a specific address (the first parameter) exists in second parameter of type map, if it doesn't the function returns the third parameter passed to it;
+```
+private function lookupByAddress(k : address, m, v) =
+  	switch(Map.lookup(k, m))
+	    None    => v
+	    Some(x) => x
+```
+
+### Implementation of the fungible token
+
+We will write a number of functions which will allow us to find out the balances of accounts as well as to transfer them from one account to another under varying conditions. These functions are described below:
+
+- **balanceOf** - the function provides the number of tokens held by a given address. ***Note*** that anyone can query any address’ balance, as all data on the blockchain is public;
+
+```
+public function balanceOf(who: address) : int = lookupByAddress(who, state._balances, 0)
+```
+
+- **totalSupply** - the return value equals the sum of all balances;
+```
+public function totalSupply() : int = state._totalSupply
+```
+
+- **transfer** - the function transfers a number of tokens directly from the caller to another address;
+```
+public stateful function transfer(to: address, value: int) : bool =
+    _transfer(Call.caller, to, value)
+    
+private stateful function _transfer(from: address, to: address, value: int) : bool =
+    require(value > 0, "Value is sub zero")
+    require(value =< balanceOf(from), "Not enough balance")
+    require(to != #0, "Invalid address")
```

Comment:
could you explain why `#0` is an invalid address? If I intent token burn it should be technical valid, right?

Reply:
@thepiwo I talked with the developer who has written FungibleToken.aes. He thinks if there is a need to burn tokens, that could be achieved by calling a burn() function which can remove the funds from the user balance and from the total supply, not by calling transfer to zero address.

Reply:
Okay, then let's keep it like this, a comment could mention that we want to prevent burning, because technically it is not invalid